src/agentlab/experiments/exp_utils.py

```diff
@@ -19,8 +22,143 @@
+
+
+def _episode_timeout(exp_arg: ExpArgs, avg_step_timeout=30):
+    """Some logic to determine the episode timeout."""
+    max_steps = getattr(exp_arg.env_args, "max_steps", None)
+    if max_steps is None:
+        episode_timeout_global = 10 * 60 * 60  # 10 hours
+    else:
+        episode_timeout_global = exp_arg.env_args.max_steps * avg_step_timeout
+
+    episode_timeout_exp = getattr(exp_arg, "episode_timeout", episode_timeout_global)
+
+    return min(episode_timeout_global, episode_timeout_exp)
+
+
+@contextmanager
+def timeout_manager(seconds: int = None):
+    """Context manager to handle timeouts."""
+
+    # Check if we're on Windows
+    if seconds is None or sys.platform == "win32":
+        try:
+            yield
+        finally:
+            pass
+    else:
+        seconds = max(1, int(seconds))
+
+        def handler(signum, frame):
+            print("before raising timeout")
+            raise TimeoutError(f"Operation timed out after {seconds} seconds")
+
+        print(f"Setting timeout to {seconds} seconds.")
+        # Register the signal handler
+        previous_handler = signal.signal(signal.SIGALRM, handler)
+        # Set the alarm
+        signal.alarm(seconds)
+
+        try:
+            yield
+        finally:
+            # Cleanup: cancel alarm and restore previous handler
+            signal.alarm(0)
+            signal.signal(signal.SIGALRM, previous_handler)
+
+
+def add_dependencies(exp_args_list: list[ExpArgs], task_dependencies: dict[str, list[str]] = None):
+    """Add dependencies to a list of ExpArgs.
+
+    Args:
+        exp_args_list: list[ExpArgs]
+            A list of experiments to run.
+        task_dependencies: dict
+            A dictionary mapping task names to a list of task names that they
+            depend on. If None or empty, no dependencies are added.
+
+    Returns:
+        list[ExpArgs]
+            The modified exp_args_list with dependencies added.
+    """
+
+    if task_dependencies is None or all([len(dep) == 0 for dep in task_dependencies.values()]):
+        # nothing to be done
+        return exp_args_list
+
+    exp_args_map = {exp_args.env_args.task_name: exp_args for exp_args in exp_args_list}
+    if len(exp_args_map) != len(exp_args_list):
+        raise ValueError(
+            (
+                "Task names are not unique in exp_args_map, "
+                "you can't run multiple seeds with task dependencies."
+            )
+        )
+
+    for task_name in exp_args_map.keys():
+        if task_name not in task_dependencies:
+            raise ValueError(f"Task {task_name} is missing from task_dependencies")
+
+    # turn dependencies from task names to exp_ids
+    for task_name, exp_args in exp_args_map.items():
+
+        exp_args.depends_on = tuple(
+            exp_args_map[dep_name].exp_id
+            for dep_name in task_dependencies[task_name]
+            if dep_name in exp_args_map  # ignore dependencies that are not to be run
```

Comment:
You should not ignore but raise an error in that case. If your graph is A -> B -> C but you re-run A and C, then C should depend on A by propagation.